File: sdk/android/core/src/main/kotlin/com/iapstack/core/IapStackClient.kt
package com.iapstack.core

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.TimeoutCancellationException
import kotlinx.coroutines.delay
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withTimeout
import okhttp3.Call
import okhttp3.Callback
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody
import okio.BufferedSink
import okhttp3.Response
import okhttp3.ResponseBody
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.InterruptedIOException
import java.net.SocketTimeoutException
import java.util.concurrent.TimeUnit
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlin.math.abs
import kotlin.random.Random
import kotlin.time.Duration

private const val SDK_VERSION = "0.1.0-dev.1"
private val JSON_MEDIA_TYPE = "application/json".toMediaType()

/**
 * Provider-neutral SDK transport for IAPStack v1.
 */
class IapStackClient(
  private val config: IapStackConfig,
  httpClient: OkHttpClient? = null,
) {
  private val gson = Gson()
  private val random = Random(System.nanoTime())
  private val closeClient: Boolean
  private val client: OkHttpClient
  private val mapType = object : TypeToken<Map<String, Any?>>() {}.type

  init {
    config.validate()
    closeClient = httpClient == null
    val timeoutMillis = config.timeout.inWholeMilliseconds
    client = (httpClient ?: OkHttpClient()).newBuilder()
      .callTimeout(timeoutMillis, TimeUnit.MILLISECONDS)
      .connectTimeout(timeoutMillis, TimeUnit.MILLISECONDS)
      .readTimeout(timeoutMillis, TimeUnit.MILLISECONDS)
      .writeTimeout(timeoutMillis, TimeUnit.MILLISECONDS)
      .build()
  }

  /**
   * Sends one purchase verification request and returns the entitlement projection.
   */
  suspend fun verifyPurchase(
    purchase: PurchaseSubmission,
    requestId: String? = null,
  ): VerificationResult {
    validatePurchase(purchase)
    val json = request(
      method = "POST",
      path = listOf("v1", "applications", config.applicationId, "purchases:verify"),
      body = purchase.toJson(),
      requestId = requestId,
    )
    return decode { VerificationResult.fromJson(json) }
  }

  /**
   * Verifies a bounded set of purchased rows.
   */
  suspend fun restorePurchases(
    purchases: List<PurchaseSubmission>,
    requestId: String? = null,
  ): RestoreResult {
    if (purchases.isEmpty() || purchases.size > 100) {
      throw IllegalArgumentException("purchases must contain between 1 and 100 items")
    }
    purchases.forEach(::validatePurchase)
    val json = request(
      method = "POST",
      path = listOf("v1", "applications", config.applicationId, "purchases:restore"),
      body = mapOf("purchases" to purchases.map(PurchaseSubmission::toJson)),
      requestId = requestId,
    )
    return decode { RestoreResult.fromJson(json) }
  }

  /**
   * Loads the current entitlement snapshot without provider interaction.
   */
  suspend fun getEntitlements(
    externalCustomerId: String,
    requestId: String? = null,
  ): EntitlementSnapshot {
    if (externalCustomerId.isBlank()) {
      throw IllegalArgumentException("externalCustomerId must not be empty")
    }
    val json = request(
      method = "GET",
      path = listOf(
        "v1",
        "applications",
        config.applicationId,
        "customers",
        externalCustomerId,
        "entitlements",
      ),
      requestId = requestId,
    )
    return decode { EntitlementSnapshot.fromJson(json) }
  }

  /**
   * Releases a custom HTTP transport when the SDK created it.
   */
  fun close() {
    if (closeClient) {
      client.connectionPool.evictAll()
      client.dispatcher.executorService.shutdown()
    }
  }

  private fun <T> decode(block: () -> T): T =
    try {
      block()
    } catch (error: IapStackProtocolException) {
      throw error
    } catch (error: RuntimeException) {
      throw IapStackProtocolException("IAPStack response did not match the v1 contract", error)
    }

  private suspend fun request(
    method: String,
    path: List<String>,
    body: Map<String, Any?>? = null,
    requestId: String? = null,
  ): Map<String, Any?> {
    val normalizedRequestId = requestId?.trim()?.ifEmpty { null }
    if (config.timeout <= Duration.ZERO) {
      throw IllegalArgumentException("timeout must be positive")
    }
    if (body != null && body.isEmpty()) {
      throw IllegalArgumentException("body must not be empty")
    }

    for (attempt in 1..config.retryPolicy.maxAttempts) {
      val request = buildRequest(method, path, body, normalizedRequestId)
      try {
        val bodyText = withTimeout(config.timeout) {
          execute(request)
        }
        return decodeObject(bodyText)
      } catch (error: IapStackApiException) {
        if (attempt >= config.retryPolicy.maxAttempts || !error.retryable) {
          throw error
        }
        delay(config.retryPolicy.delayAfter(attempt, jitter()))
      } catch (error: TimeoutCancellationException) {
        if (attempt >= config.retryPolicy.maxAttempts) {
          throw IapStackTimeoutException("IAPStack request timed out", error)
        }
        delay(config.retryPolicy.delayAfter(attempt, jitter()))
      } catch (error: CancellationException) {
        throw error
      } catch (error: SocketTimeoutException) {
        if (attempt >= config.retryPolicy.maxAttempts) {
          throw IapStackTimeoutException("IAPStack request timed out", error)
        }
        delay(config.retryPolicy.delayAfter(attempt, jitter()))
      } catch (error: InterruptedIOException) {
        if (attempt >= config.retryPolicy.maxAttempts) {
          throw IapStackTimeoutException("IAPStack request timed out", error)
        }
        delay(config.retryPolicy.delayAfter(attempt, jitter()))
      } catch (error: IOException) {
        if (attempt >= config.retryPolicy.maxAttempts) {
          throw IapStackTransportException(
            "IAPStack request failed before a response was received",
            error,
          )
        }
        delay(config.retryPolicy.delayAfter(attempt, jitter()))
      } catch (error: IapStackProtocolException) {
        throw error
      }
    }
    throw IapStackTransportException("IAPStack request exhausted the retry policy")
  }

  private suspend fun execute(request: Request): String {
    val call = client.newCall(request)
    val response = suspendCancellableCoroutine<Response> { cont ->
      cont.invokeOnCancellation { call.cancel() }
      call.enqueue(object : Callback {
        override fun onFailure(call: Call, e: IOException) {
          if (cont.isActive) {
            cont.resumeWithException(e)
          }
        }

        override fun onResponse(call: Call, response: Response) {
          if (!cont.isActive) {
            response.close()
            return
          }
          cont.resume(response)
        }
      })
    }
    return response.use { resp ->
      val responseBody = resp.body
        ?: throw IapStackProtocolException("IAPStack response was empty")
      val rawBody = collectResponseBody(responseBody)
      if (resp.isSuccessful) {
        rawBody
      } else {
        throw apiErrorFromResponse(resp.code, rawBody)
      }
    }
  }

  private fun jitter(): Double = abs(random.nextDouble())

  private fun buildRequest(
    method: String,
    path: List<String>,
    body: Map<String, Any?>?,
    requestId: String?,
  ): Request {
    val urlBuilder = config.baseUri.toString().toHttpUrl().newBuilder()
    for (segment in path) {
      urlBuilder.addPathSegment(segment)
    }
    val requestBuilder = Request.Builder()
      .url(urlBuilder.build())
      .addHeader("Accept", "application/json")
      .addHeader("Authorization", "Bearer ${config.customerToken}")
      .addHeader("X-IAPStack-SDK", "android-kotlin/$SDK_VERSION")

    if (requestId != null && requestId.isBlank()) {
      throw IllegalArgumentException("requestId cannot be blank")
    }
    requestId?.let {
      requestBuilder.addHeader("X-Request-ID", it)
    }

    val requestBody = body?.let {
      val rawBody = gson.toJson(it)
      requestBuilder.addHeader("Content-Type", "application/json")
      jsonRequestBody(rawBody)
    }

    return requestBuilder
      .method(method, requestBody)
      .build()
  }

  private fun jsonRequestBody(rawBody: String): RequestBody =
    object : RequestBody() {
      override fun contentType() = JSON_MEDIA_TYPE

      override fun contentLength() = rawBody.toByteArray(Charsets.UTF_8).size.toLong()

      override fun writeTo(sink: BufferedSink) {
        sink.writeString(rawBody, Charsets.UTF_8)
      }
    }

  private fun collectResponseBody(responseBody: ResponseBody): String {
    val raw = ByteArrayOutputStream()
    responseBody.byteStream().use { stream ->
      val buffer = ByteArray(4096)
      var total = 0
      while (true) {
        val read = stream.read(buffer)
        if (read < 0) {
          break
        }
        total += read
        if (total > config.maxResponseBytes) {
          throw IapStackProtocolException("IAPStack response exceeded the configured size limit")
        }
        raw.write(buffer, 0, read)
      }
    }
    return raw.toString(Charsets.UTF_8)
  }

  private fun decodeObject(value: String): Map<String, Any?> =
    try {
      asObjectMap(gson.fromJson(value, mapType))
        ?: throw IapStackProtocolException("IAPStack response was not a JSON object")
    } catch (error: IapStackProtocolException) {
      throw error
    } catch (error: RuntimeException) {
      throw IapStackProtocolException("IAPStack returned an invalid JSON object", error)
    }

  private fun apiErrorFromResponse(
    statusCode: Int,
    responseBody: String,
  ): IapStackApiException {
    val parsed = try {
      decodeObject(responseBody)
    } catch (_: Exception) {
      return IapStackApiException(
        statusCode = statusCode,
        code = "http_error",
        requestId = null,
        retryable = statusCode == 429 || statusCode >= 500,
        message = "IAPStack returned an unsuccessful response",
      )
    }
    val errorObject = parsed["error"] as? Map<*, *>
    val code = (errorObject?.get("code") as? String)?.ifBlank { "http_error" } ?: "http_error"
    val message = (errorObject?.get("message") as? String)?.ifBlank {
      "IAPStack returned an unsuccessful response"
    } ?: "IAPStack returned an unsuccessful response"
    val requestId = errorObject?.get("request_id") as? String
    return IapStackApiException(
      statusCode = statusCode,
      code = code,
      requestId = requestId,
      retryable = statusCode == 429 || statusCode >= 500,
      message = message,
    )
  }

  private fun validatePurchase(purchase: PurchaseSubmission) {
    if (purchase.externalCustomerId.isBlank()) {
      throw IllegalArgumentException("externalCustomerId must not be empty")
    }
    if (purchase.claimedProducts.isEmpty() || purchase.evidence.isEmpty()) {
      throw IllegalArgumentException("purchase must contain claimed products and evidence")
    }
    if (purchase.claimedProducts.any(String::isBlank)) {
      throw IllegalArgumentException("claimed products must not be empty")
    }
    if (purchase.claimedProducts.size != purchase.claimedProducts.toSet().size) {
      throw IllegalArgumentException("claimed products must be unique")
    }
  }
}

File: sdk/ios/Sources/IAPStackApple/IAPStackClient.swift
import Foundation

private let sdkVersion = "0.1.0-dev.1"

/// Provider-neutral HTTP client for the IAPStack v1 API.
public final class IAPStackClient {
  /// Creates a client with optional custom URLSession for testing.
  public init(config: IAPStackConfig, session: URLSession = .shared) throws {
    try config.validate()
    self.config = config
    self.session = session
    self.ownsSession = session !== URLSession.shared
  }

  private let config: IAPStackConfig
  private let session: URLSession
  private let ownsSession: Bool

  /// Verifies one signed purchase evidence row.
  public func verifyPurchase(_ purchase: PurchaseSubmission, requestId: String? = nil) async throws -> VerificationResult {
    let body = purchase.toDictionary()
    let json = try await request(
      method: "POST",
      path: ["v1", "applications", config.applicationId, "purchases:verify"],
      body: body,
      requestId: requestId,
    )
    return try VerificationResult.from(json)
  }

  /// Verifies a bounded batch of signed purchases.
  public func restorePurchases(
    _ purchases: [PurchaseSubmission],
    requestId: String? = nil,
  ) async throws -> RestoreResult {
    if purchases.isEmpty || purchases.count > 100 {
      throw IAPStackSDKError.configurationError(
        message: "purchases must contain between 1 and 100 items",
      )
    }
    let json = try await request(
      method: "POST",
      path: ["v1", "applications", config.applicationId, "purchases:restore"],
      body: ["purchases": purchases.map { $0.toDictionary() }],
      requestId: requestId,
    )
    return try RestoreResult.from(json)
  }

  /// Loads the current entitlement snapshot for one external customer.
  public func getEntitlements(_ externalCustomerId: String, requestId: String? = nil) async throws
    -> EntitlementSnapshot {
    if externalCustomerId.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
      throw IAPStackSDKError.configurationError(
        message: "externalCustomerId must not be empty",
      )
    }
    let json = try await request(
      method: "GET",
      path: [
        "v1",
        "applications",
        config.applicationId,
        "customers",
        externalCustomerId,
        "entitlements",
      ],
      body: nil,
      requestId: requestId,
    )
    return try EntitlementSnapshot.from(json)
  }

  /// Releases the owned `URLSession`.
  public func close() {
    if ownsSession {
      session.invalidateAndCancel()
    }
  }

  private func request(
    method: String,
    path: [String],
    body: [String: Any]?,
    requestId: String?,
  ) async throws -> [String: Any] {
    let normalizedRequestId = requestId?.trimmingCharacters(in: .whitespacesAndNewlines)
    if let normalized = normalizedRequestId, normalized.isEmpty {
      throw IAPStackSDKError.configurationError(message: "requestId cannot be empty")
    }

    var attempt = 1
    while attempt <= config.retryPolicy.maxAttempts {
      do {
        return try await requestAttempt(
          method: method,
          path: path,
          body: body,
          requestId: normalizedRequestId,
        )
      } catch {
        if shouldRetry(error: error, attempt: attempt) {
          let delay = config.retryPolicy.delayAfter(
            attempt: attempt,
            randomValue: Double.random(in: 0 ... 1),
          )
          if delay > 0 {
            try await Task.sleep(nanoseconds: UInt64(delay * 1_000_000_000))
          }
          attempt += 1
          continue
        }
        throw error
      }
    }
    throw IAPStackSDKError.transportError(message: "IAPStack request exhausted retry policy")
  }

  private func requestAttempt(
    method: String,
    path: [String],
    body: [String: Any]?,
    requestId: String?,
  ) async throws -> [String: Any] {
    let (data, response) = try await withTimeout {
      try await self.send(method: method, path: path, body: body, requestId: requestId)
    }
    if data.count > config.maxResponseBytes {
      throw IAPStackSDKError.protocolError(message: "IAPStack response exceeded maxResponseBytes")
    }
    let json = try decodeJSONObject(from: data)
    if (200 ... 299).contains(response.statusCode) {
      return json
    }
    throw parseApiError(statusCode: response.statusCode, response: response, body: json)
  }

  private func send(
    method: String,
    path: [String],
    body: [String: Any]?,
    requestId: String?,
  ) async throws -> (Data, HTTPURLResponse) {
    let url = append(path: path)
    var request = URLRequest(url: url)
    request.httpMethod = method
    request.setValue("application/json", forHTTPHeaderField: "Accept")
    request.setValue("Bearer \(config.customerToken)", forHTTPHeaderField: "Authorization")
    request.setValue("ios-swift/\(sdkVersion)", forHTTPHeaderField: "X-IAPStack-SDK")
    if let requestId {
      request.setValue(requestId, forHTTPHeaderField: "X-Request-ID")
    }

    if let body {
      if body.isEmpty {
        throw IAPStackSDKError.configurationError(message: "request body must not be empty")
      }
      request.httpBody = try JSONSerialization.data(withJSONObject: body)
      request.setValue("application/json", forHTTPHeaderField: "Content-Type")
    }

    let (data, response) = try await session.data(for: request)
    guard let httpResponse = response as? HTTPURLResponse else {
      throw IAPStackSDKError.protocolError(message: "non-HTTP response from server")
    }
    return (data, httpResponse)
  }

  private func withTimeout<T>(_ operation: @escaping () async throws -> T) async throws -> T {
    let timeout = max(1, UInt64(config.timeout * 1_000_000_000))
    do {
      return try await withThrowingTaskGroup(of: TimeoutRace<T>.self) { group in
        group.addTask { .value(try await operation()) }
        group.addTask {
          try await Task.sleep(nanoseconds: timeout)
          return .timeout
        }
        defer { group.cancelAll() }
        switch try await group.next() {
        case .value(let value):
          return value
        case .timeout:
          throw IAPStackSDKError.timeoutError(message: "IAPStack request timed out")
        case nil:
          throw IAPStackSDKError.transportError(message: "IAPStack request did not execute")
        }
      }
    } catch is CancellationError {
      throw IAPStackSDKError.timeoutError(message: "IAPStack request timed out")
    } catch {
      if case IAPStackSDKError.timeoutError = error {
        throw error
      }
      if let urlError = error as? URLError,
        urlError.code == .timedOut || urlError.code == .cancelled
      {
        throw IAPStackSDKError.timeoutError(
          message: "IAPStack request timed out",
          cause: error,
        )
      }
      if let urlError = error as? URLError {
        throw IAPStackSDKError.transportError(
          message: "IAPStack request failed before a response was received",
          cause: urlError,
        )
      }
      throw error
    }
  }

  private func shouldRetry(error: Error, attempt: Int) -> Bool {
    guard attempt < config.retryPolicy.maxAttempts else {
      return false
    }
    if error is CancellationError {
      return false
    }
    guard let sdkError = error as? IAPStackSDKError else {
      return error is URLError
    }
    return sdkError.isRetryable
  }

  private func parseApiError(
    statusCode: Int,
    response: HTTPURLResponse,
    body: [String: Any],
  ) -> IAPStackSDKError {
    var code = "http_error"
    var message = "IAPStack returned an unsuccessful response"
    var requestId: String?
    if let errorObject = body["error"] as? [String: Any] {
      if let parsedCode = errorObject["code"] as? String, !parsedCode.isEmpty {
        code = parsedCode
      }
      if let parsedMessage = errorObject["message"] as? String, !parsedMessage.isEmpty {
        message = parsedMessage
      }
      requestId = errorObject["request_id"] as? String
    }
    let headerRequestId = response.allHeaderFields.first(
      where: { ($0.key as? String)?.lowercased() == "x-request-id" },
    )?.value as? String
    return IAPStackSDKError.apiError(
      statusCode: statusCode,
      code: code,
      message: message,
      requestId: requestId ?? headerRequestId,
      retryable: statusCode == 429 || statusCode >= 500,
    )
  }

  private func decodeJSONObject(from data: Data) throws -> [String: Any] {
    if data.isEmpty {
      throw IAPStackSDKError.protocolError(message: "empty response body")
    }
    let object: Any
    do {
      object = try JSONSerialization.jsonObject(with: data)
    } catch {
      throw IAPStackSDKError.protocolError(message: "IAPStack returned invalid JSON", cause: error)
    }
    guard let dictionary = object as? [String: Any] else {
      throw IAPStackSDKError.protocolError(message: "IAPStack response was not a JSON object")
    }
    return dictionary
  }

  private func append(path segments: [String]) -> URL {
    var components = URLComponents(url: config.baseUri, resolvingAgainstBaseURL: false)
      ?? URLComponents()
    var path = components.percentEncodedPath
    if path.isEmpty {
      path = "/"
    }
    if !path.hasSuffix("/") {
      path += "/"
    }
    var allowed = CharacterSet.urlPathAllowed
    allowed.remove(charactersIn: "/")
    let encoded = segments.map { segment in
      segment.addingPercentEncoding(withAllowedCharacters: allowed) ?? segment
    }
    path += encoded.joined(separator: "/")
    components.percentEncodedPath = path
    return components.url ?? config.baseUri
  }
}

private enum TimeoutRace<Value> {
  case value(Value)
  case timeout
}
